Error: Cannot find module 'joi'

## Description

Gatsby immediately throws an exception. 

### Steps to reproduce

`gatsby`

### Expected result

Usage message, etc.

### Actual result
```
C:\Users\_USER_\AppData\Roaming\npm\node_modules\gatsby-cli\node_modules\yoga-layout-prebuilt\yoga-layout\build\Release\nbind.js:53
        throw ex;
        ^

Error: Cannot find module 'joi'
    at Function.Module._resolveFilename (module.js:548:15)
    at Function.Module._load (module.js:475:25)
    at Module.require (module.js:597:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (C:\Users\johnp\AppData\Roaming\npm\node_modules\gatsby-cli\lib\structured-errors\construct-error.js:3:13)
    at Module._compile (module.js:653:30)
    at Object.Module._extensions..js (module.js:664:10)
    at Module.load (module.js:566:32)
    at tryModuleLoad (module.js:506:12)
    at Function.Module._load (module.js:498:3)
```
### Environment

gatsby command doesn't work period. 
npm version 6.9.2, i've tried uninstalling and reinstalling gatsby etc.

Ditto on node 11.12:
```
/.nvm/versions/node/v11.12.0/lib/node_modules/gatsby-cli/node_modules/yoga-layout-prebuilt/yoga-layout/build/Release/nbind.js:53
        throw ex;
        ^

Error: Cannot find module 'joi'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:649:15)
    at Function.Module._load (internal/modules/cjs/loader.js:575:25)
    at Module.require (internal/modules/cjs/loader.js:705:19)
    at require (internal/modules/cjs/helpers.js:14:16)
    at Object.<anonymous> (/Users/foo/.nvm/versions/node/v11.12.0/lib/node_modules/gatsby-cli/lib/structured-errors/construct-error.js:3:13)
    at Module._compile (internal/modules/cjs/loader.js:799:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:810:10)
    at Module.load (internal/modules/cjs/loader.js:666:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:606:12)
    at Function.Module._load (internal/modules/cjs/loader.js:598:3)
```

Just updated the cli .... it's broken:

```
.nvm/versions/node/v11.15.0/lib/node_modules/gatsby-cli/node_modules/yoga-layout-prebuilt/yoga-layout/build/Release/nbind.js:53
        throw ex;
        ^

Error: Cannot find module 'joi'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:668:15)
```

/home/chun/.nvm/versions/node/v10.16.0/lib/node_modules/gatsby/node_modules/yoga-layout-prebuilt/yoga-layout/build/Release/nbind.js:53
        throw ex;
        ^

Error: Cannot find module 'joi'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)

Same issue 

`Error: Cannot find module 'joi'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:582:15)
    at Function.Module._load (internal/modules/cjs/loader.js:508:25)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:22:18)
    at Object.<anonymous> (/usr/local/lib/node_modules/gatsby-cli/lib/structured-errors/construct-error.js:3:13)
    at Module._compile (internal/modules/cjs/loader.js:701:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:712:10)
    at Module.load (internal/modules/cjs/loader.js:600:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:539:12)
    at Function.Module._load (internal/modules/cjs/loader.js:531:3)
Genisys:forwardme volkan$ gatsby build
/usr/local/lib/node_modules/gatsby-cli/node_modules/yoga-layout-prebuilt/yoga-layout/build/Release/nbind.js:53
        throw ex;`

--- a/packages/gatsby-cli/package.json
+++ b/packages/gatsby-cli/package.json
@@ -12,6 +12,7 @@
   "dependencies": {
     "@babel/code-frame": "^7.0.0",
     "@babel/runtime": "^7.0.0",
+    "@hapi/joi": "^15.1.0",
     "better-opn": "^0.1.4",
     "bluebird": "^3.5.0",
     "chalk": "^2.4.2",

--- a/packages/gatsby-cli/src/structured-errors/__tests__/construct-error.js
+++ b/packages/gatsby-cli/src/structured-errors/__tests__/construct-error.js
@@ -1,0 +1,37 @@
+const constructError = require(`../construct-error`)
+
+let log
+let processExit
+beforeEach(() => {
+  log = jest.spyOn(console, `log`).mockImplementation(() => {})
+  processExit = jest.spyOn(process, `exit`).mockImplementation(() => {})
+
+  log.mockReset()
+  processExit.mockReset()
+})
+
+afterAll(() => {
+  console.log.mockClear()
+  process.exit.mockClear()
+})
+
+test(`it exits on invalid error schema`, () => {
+  constructError({ details: { context: {}, lol: `invalid` } })
+
+  expect(processExit).toHaveBeenCalledWith(1)
+})
+
+test(`it logs error on invalid schema`, () => {
+  constructError({ details: { context: {}, lol: `invalid` } })
+
+  expect(log).toHaveBeenCalledWith(
+    `Failed to validate error`,
+    expect.any(Object)
+  )
+})
+
+test(`it passes through on valid error schema`, () => {
+  constructError({ details: { context: {} } })
+
+  expect(log).not.toHaveBeenCalled()
+})

--- a/packages/gatsby-cli/src/structured-errors/__tests__/error-map.js
+++ b/packages/gatsby-cli/src/structured-errors/__tests__/error-map.js
@@ -1,0 +1,23 @@
+const { errorMap, defaultError } = require(`../error-map`)
+
+test(`it defaults to generic error`, () => {
+  expect(defaultError).toEqual(
+    expect.objectContaining({
+      level: `ERROR`,
+    })
+  )
+
+  expect(defaultError.text({})).toEqual(`There was an error`)
+})
+
+test(`it supports structured lookups`, () => {
+  const error = errorMap[`95312`]
+
+  expect(error).toEqual(
+    expect.objectContaining({
+      text: expect.any(Function),
+      docsUrl: `https://gatsby.dev/debug-html`,
+      level: `ERROR`,
+    })
+  )
+})

--- a/packages/gatsby-cli/src/structured-errors/__tests__/error-schema.js
+++ b/packages/gatsby-cli/src/structured-errors/__tests__/error-schema.js
@@ -1,0 +1,13 @@
+const schema = require(`../error-schema`)
+
+test(`throws invalid on an invalid error`, () => {
+  expect(schema.validate({ lol: `true` })).rejects.toBeDefined()
+})
+
+test(`does not throw on a valid schema`, () => {
+  expect(
+    schema.validate({
+      context: {},
+    })
+  ).resolves.toEqual(expect.any(Object))
+})

--- a/packages/gatsby-cli/src/structured-errors/construct-error.js
+++ b/packages/gatsby-cli/src/structured-errors/construct-error.js
@@ -1,4 +1,4 @@
-const Joi = require(`joi`)
+const Joi = require(`@hapi/joi`)
 const stackTrace = require(`stack-trace`)
 const errorSchema = require(`./error-schema`)
 const { errorMap, defaultError } = require(`./error-map`)

--- a/packages/gatsby-cli/src/structured-errors/error-schema.js
+++ b/packages/gatsby-cli/src/structured-errors/error-schema.js
@@ -1,4 +1,4 @@
-const Joi = require(`joi`)
+const Joi = require(`@hapi/joi`)
 
 const Position = Joi.object().keys({
   line: Joi.number(),

